txerr/doc.go

```diff
@@ -0,0 +1,168 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Package txerr provides common functions for querying and formatting errors
+// generated by txfile.
+//
+// Creating errors in txfile for use with txerr must follow some common
+// conventions, so to guarantee some level consistency among packages.
+//
+// Package txerr expects errors to implement methods for querying error
+// internals.  This allows us to treat errors as interfaces, while having
+// package specific implementations.  A common set of recommended methods is
+// documented with the Error type.
+//
+// The Error type is designed to rely on builtin types only. No implementation
+// must be forced to use a type define in txerr or any other common package.
+//
+// Errors are treated as a tree of individual error values. Each node in the
+// tree can have 0, 1 or N children. Errors capturing a single cause only must
+// implement `Cause() error`. Errors capturing multiple casues must implement
+// `Causes() []error`. Leaf-Nodes either do not implement Cause,Causes or must
```

Comment:
Why not be stricter with Leaf-Nodes by accepting only one "true" way to distingush them from non-leaf nodes? E.g saying that Leaf-nodes should return nil and not implementing `Causes` and/or `Cause` is not acceptable. Stricter distinction would lead to more straightforward code.